llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -411,6 +417,21 @@ std::string saveModuleIR(Module &M, int I, StringRef Suff) {
   return OutFilename;
 }
 
+bool isImportedFunction(const Function &F) {
+  // Functions with definitions are not imported
```

Comment:
We are not checking if a function has actually been called, but if `StripDeadPrototypesPass` hadn't been applied to a module we may have some unused symbols recorded as dependencies. The latter could happen under `-O0` for example, or as a result of some other optimization/transformation done by `sycl-post-link`.

I would suggest to add an early exit here if a function has no uses.

Reply:
Dead prototypes are removed in module splitting.  In ModuleSplitter.cpp ModuleDesc::cleanup()

```
  MPM.addPass(GlobalDCEPass());           // Delete unreachable globals.
  MPM.addPass(StripDeadDebugInfoPass());  // Remove dead debug info.
  MPM.addPass(StripDeadPrototypesPass()); // Remove dead func decls.
```
Both GlobalDCEPass and StripDeadPrototypesPass needed to be commented out in order to have a dead prototype reach `isImportedFunction`.

Do we still need to check for dead prototypes in `isImportedFunction`?
Should we remove the call to StripDeadPrototypesPass in sycl-post-link.cpp?

Reply:
Added assertion.